dao-voting/src/main.sw

```diff
@@ -0,0 +1,295 @@
+contract;
+
+dep abi;
+dep data_structures;
+dep errors;
+dep utils;
+
+use std::{
+    address::Address,
+    assert::require,
+    block::height,
+    chain::auth::{AuthError, msg_sender},
+    context::{call_frames::msg_asset_id, msg_amount, this_balance},
+    contract_id::ContractId,
+    identity::Identity,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+    token::{force_transfer_to_contract, transfer_to_output}
+};
+
+use abi::DaoVoting;
+use data_structures::{CallData, Proposal};
+use errors::Error;
+use utils::sender_identity;
+
+storage {
+    gov_token: ContractId,
+    voting_period: u64,
+    approval_percentage: u64,
+    proposals: StorageMap<u64,
+    Proposal>, proposal_count: u64,
+    // The amount of governance tokens a user has deposited
+    balances: StorageMap<Identity,
+    u64>, // The amount of votes a user has used on a proposal
+    votes: StorageMap<(Identity,
+    u64), u64>, state: u64,
+}
+
+impl DaoVoting for Contract {
+    /// Initialize the dao with the governance token, voting parameters, and the proposal.
+    ///
+    /// # Parameters
+    ///
+    /// gov_token - contract id of the token to use to vote on governance proposals
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor is called more than once
+    /// - The voting period is not greater than 0
+    /// - The approval percentage is not greater than 0
+    #[storage(read, write)]fn constructor(gov_token: ContractId) -> bool {
+        require(storage.state == 0, Error::CannotReinitialize);
+
+        storage.gov_token = gov_token;
+        storage.proposal_count = 0;
+        storage.state = 1;
+
+        true
+    }
+
+    /// Add proposal to be voted on
+    ///
+    /// # Parameters
+    ///
+    /// voting_period - the number of blocks during which a proposal can be voted on
+    /// approval_percentage - the percentage of yes votes a proposal needs to be executed
+    /// proposal_data - transaction data to be executed if proposal is approved
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called to initialize
+    /// - The voting period is 0
+    /// - The approval percentage is 0
+    /// - The approval percentage is above 100
+    #[storage(read, write)]fn add_proposal(voting_period: u64, approval_percentage: u64, proposal_data: CallData) -> bool {
+        require(storage.state == 1, Error::NotInitialized);
+        require(voting_period > 0, Error::PeriodCannotBeZero);
```

Comment:
Is there a particular reason to order the expression this way?

Reply:
Consistency with other apps. I also find syntax like `x >= y` to be strange rather than `y <= x`. This is how I've seen math notation be taught and also coding so I am assuming that it is far more common than the alternative (which apparently people use).